File: 7.二叉树/18.从中序与后序遍历序列构造二叉树/106.从中序与后序遍历序列构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=106 lang=cpp
 *
 * [106] 从中序与后序遍历序列构造二叉树
 */
#include <bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (! inorder.size() || ! postorder.size()) return nullptr;
        return dfs(inorder, postorder, 0, inorder.size() - 1, 0, postorder.size() - 1);
    }
private:
    TreeNode* dfs(vector<int>& in, vector<int>& post, int in_s, int in_e, int po_s, int po_e)
    {
        if (po_s > po_e) return nullptr;
        int tmp = post[po_e];
        TreeNode* root = new TreeNode(tmp);
        if (po_s == po_e) return root;
        int mid;
        for (mid = in_s; mid <= in_e; mid ++)
        {
            if (in[mid] == tmp) break;
        }
        int lin_s = in_s, lin_e = mid - 1;
        int rin_s = mid + 1, rin_e = in_e;
        po_e --;
        int lpo_s = po_s, lpo_e = po_s + lin_e - lin_s;
        int rpo_s = lpo_e + 1, rpo_e = po_e;
        root->left = dfs(in, post, lin_s, lin_e, lpo_s, lpo_e);
        root->right = dfs(in, post, rin_s, rin_e, rpo_s, rpo_e);
        return root;
    }
};
// @lc code=end

class Solution {
    public:
        TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
            if (! inorder.size() || ! postorder.size()) return nullptr;
            return dfs(inorder, postorder);
        }
    private:
        TreeNode* dfs(vector<int>& in, vector<int>& post)
        {
            if (! post.size()) return nullptr;
            int tmp = post[post.size() - 1];
            TreeNode* root = new TreeNode(tmp);
            if (post.size() == 1) return root;
            int mid = 0;
            for (; mid < in.size(); mid ++)
            {
                if (in[mid] == tmp) break;
            }
            vector<int> lin(in.begin(), in.begin() + mid);
            vector<int> rin(in.begin() + mid + 1, in.end());
            post.resize(post.size() - 1);
            vector<int> lpo(post.begin(), post.begin() + lin.size());
            vector<int> rpo(post.begin() + lin.size(), post.end());
            root->left = dfs(lin, lpo);
            root->right = dfs(rin, rpo);
            return root;
        }
    };

File: 面经150/9.二叉树/106.从中序与后序遍历序列构造二叉树.py
#
# @lc app=leetcode.cn id=106 lang=python3
#
# [106] 从中序与后序遍历序列构造二叉树
#
from typing import List, Optional
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def dfs(il, ir, pl, pr):
            if pl > pr: return None
            x = postorder[pr]
            cur = TreeNode(x)
            if pl == pr: return cur
            for mid in range(il, ir + 1):
                if inorder[mid] == x: break
            pr -= 1
            d = mid - il - 1
            lpl, lpr, rpl, rpr = pl, pl + d, pl + d + 1, pr
            lil, lir, ril, rir = il, mid - 1, mid + 1, ir
            cur.left = dfs(lil, lir, lpl, lpr)
            cur.right = dfs(ril, rir, rpl, rpr)
            return cur
        return dfs(0, len(inorder) - 1, 0, len(postorder) - 1)
# @lc code=end

